File: app/src/test/java/humbleactivity/app/ChainComposerTest.java
package humbleactivity.app;

import org.jmock.Expectations;
import org.jmock.States;
import org.jmock.integration.junit4.JUnitRuleMockery;
import org.jmock.lib.concurrent.Synchroniser;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import rx.Observable;
import rx.schedulers.Schedulers;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class ChainComposerTest {
    Synchroniser synchroniser = new Synchroniser();
    @Rule
    public JUnitRuleMockery mockery = new JUnitRuleMockery() {{
        setThreadingPolicy(synchroniser);
    }};
    ChainComposer.ChainComposerView view = mockery.mock(ChainComposer.ChainComposerView.class);
    EffectorService effectorService = mockery.mock(EffectorService.class);
    ChainComposer dut = new ChainComposer(effectorService, new RxScheduling(Schedulers.io(), Schedulers.immediate()));

    @Before
    public void setUp() {
        dut.attach(view);
    }

    @Test
    public void initialization() throws InterruptedException {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"));
        final List<Filter> chain = Collections.emptyList();
        final States states = mockery.states("listFilters");
        mockery.checking(new Expectations() {{
            oneOf(effectorService).listFilters(); will(returnValue(Observable.just(filters)));
            oneOf(view).setAvailableFilters(filters);
            oneOf(view).setChain(chain);
            then(states.is("called"));
        }});
        dut.initialize();
        synchroniser.waitUntil(states.is("called"));
    }

    @Test
    public void initializationFailure() throws InterruptedException {
        String errorMessage = "error";
        final States states = mockery.states("listFilters");
        mockery.checking(new Expectations() {{
            oneOf(effectorService).listFilters(); will(returnValue(Observable.error(new IOException(errorMessage))));
            oneOf(view).showErrorMessage(errorMessage);
            then(states.is("called"));
        }});
        dut.initialize();
        synchroniser.waitUntil(states.is("called"));
    }

    @Test
    public void addToChain() {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"), new Filter("Distortion"));
        mockery.checking(new Expectations() {{
            oneOf(view).setAvailableFilters(filters.subList(1, filters.size()));
            oneOf(view).setChain(filters.subList(0, 1));
        }});
        dut.availableFilters = new ArrayList<>(filters);
        dut.chain = new ArrayList<>();
        dut.addToChain(0);
    }

    @Test
    public void removeFromChain() {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"), new Filter("Distortion"));
        mockery.checking(new Expectations() {{
            oneOf(view).setAvailableFilters(filters);
            oneOf(view).setChain(filters.subList(0, 0));
        }});
        dut.availableFilters = new ArrayList<>();
        dut.availableFilters.addAll(filters.subList(0, 1));
        dut.chain = new ArrayList<>();
        dut.chain.addAll(filters.subList(1, filters.size()));
        dut.removeFromChain(0);
    }

    @Test
    public void refresh() throws InterruptedException {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"));
        final List<Filter> chain = Collections.emptyList();
        final States states = mockery.states("listFilters");
        mockery.checking(new Expectations() {{
            oneOf(effectorService).listFilters(); will(returnValue(Observable.just(filters)));
            oneOf(view).setAvailableFilters(filters);
            oneOf(view).setChain(chain);
            then(states.is("called"));
        }});
        dut.refresh();
        synchroniser.waitUntil(states.is("called"), 1000);
    }

    @Test
    public void moveUpFilter() {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"), new Filter("Distortion"));
        mockery.checking(new Expectations() {{
            oneOf(view).swapFilterInChain(1, 0);
        }});
        dut.chain = new ArrayList<>(filters);
        dut.moveUpFilter(1);
    }

    @Test
    public void moveDownFilter() {
        final List<Filter> filters = Arrays.asList(new Filter("Reverb"), new Filter("Distortion"));
        mockery.checking(new Expectations() {{
            oneOf(view).swapFilterInChain(0, 1);
        }});
        dut.chain = new ArrayList<>(filters);
        dut.moveDownFilter(0);
    }
}

File: app/src/test/java/humbleactivity/app/ui/ChainComposerTest.kt
package humbleactivity.app.ui

import humbleactivity.app.data.EffectorService
import humbleactivity.app.data.Filter
import io.reactivex.Observable
import io.reactivex.observers.TestObserver
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.kotlin.mock
import org.mockito.kotlin.stub
import java.io.IOException

@MockitoSettings
class ChainComposerTest {
    lateinit var effectorService: EffectorService
    lateinit var dut: ChainComposer
    lateinit var availablesSubscriber: TestObserver<List<Filter>>
    lateinit var chainSubscriber: TestObserver<List<Filter>>
    lateinit var chainCursorMoveSubscriber: TestObserver<Int>
    lateinit var loadErrorSubscriber: TestObserver<String>

    @BeforeEach
    fun setUp() {
        effectorService = mock()
        dut = ChainComposer(effectorService)
        availablesSubscriber = dut.availables.test()
        chainSubscriber = dut.chain.test()
        chainCursorMoveSubscriber = dut.chainCursor.test()
        loadErrorSubscriber = dut.loadError.test()
    }

    @Test
    fun initialization() {
        val filters = listOf(Filter("Reverb"))
        val chain = emptyList<Filter>()
        stub {
            on(effectorService.listFilters()).thenReturn(Observable.just(filters))
        }
        dut.initialize()
        availablesSubscriber.assertValues(emptyList(), filters)
        chainSubscriber.assertValues(emptyList(), chain)
    }

    @Test
    fun initializationError() {
        val errorMessage = "error"
        stub {
            on(effectorService.listFilters()).thenReturn(Observable.error(IOException(errorMessage)))
        }
        dut.initialize()
        availablesSubscriber.assertValues(emptyList())
        chainSubscriber.assertValues(emptyList())
        loadErrorSubscriber.assertValue(errorMessage)
    }

    @Test
    fun addToChain() {
        val filters = listOf(Filter("Reverb"), Filter("Distortion"))
        dut._backdoor.accept(ChainComposer.State(filters, emptyList()))
        dut.addToChain.accept(0)
        availablesSubscriber.assertValues(emptyList(), filters, filters.subList(1, filters.size))
        chainSubscriber.assertValues(emptyList(), emptyList(), filters.subList(0, 1))
    }

    @Test
    fun removeFromChain() {
        val availables = listOf(Filter("Reverb"))
        val chain = listOf(Filter("Distortion"))
        dut._backdoor.accept(ChainComposer.State(availables, chain))
        dut.removeFromChain.accept(0)
        availablesSubscriber.assertValues(emptyList(), availables, availables + chain)
        chainSubscriber.assertValues(emptyList(), chain, emptyList())
    }

    @Test
    fun refresh() {
        val filters = listOf(Filter("Reverb"))
        val chain = emptyList<Filter>()
        stub {
            on(effectorService.listFilters()).thenReturn(Observable.just(filters))
        }
        dut.refresh.accept(Unit)
        availablesSubscriber.assertValues(emptyList(), filters)
        chainSubscriber.assertValues(emptyList(), chain)
    }

    @Test
    fun moveUpFilter() {
        val chain = listOf(Filter("Reverb"), Filter("Distortion"))
        dut._backdoor.accept(ChainComposer.State(emptyList(), chain))
        dut.moveUp.accept(1)
        chainSubscriber.assertValues(emptyList(), chain, chain.subList(1, chain.size) + chain[0])
        chainCursorMoveSubscriber.assertValues(0)
    }

    @Test
    fun moveDownFilter() {
        val chain = listOf(Filter("Reverb"), Filter("Distortion"))
        dut._backdoor.accept(ChainComposer.State(emptyList(), chain))
        dut.moveDown.accept(0)
        chainSubscriber.assertValues(emptyList(), chain, chain.subList(1, chain.size) + chain[0])
        chainCursorMoveSubscriber.assertValues(1)
    }
}
